app_server.py

```diff
@@ -266,55 +266,88 @@ def heartbeat(self, conf_class=None):
 
 
 class URNLookupController(object):
-    """A generic controller that takes URNs as input and looks up their
-    OPDS entries.
+    """A controller for looking up OPDS entries for specific books,
+    identified in terms of their Identifier URNs.
     """
 
-    UNRECOGNIZED_IDENTIFIER = "This work is not in the collection."
-    WORK_NOT_PRESENTATION_READY = "Work created but not yet presentation-ready."
-    WORK_NOT_CREATED = "Identifier resolved but work not yet created."
-
     def __init__(self, _db):
+        """Constructor.
+
+        :param _db: A database connection.
+        """
         self._db = _db
-        self.works = []
-        self.precomposed_entries = []
-        self.unresolved_identifiers = []
 
     def work_lookup(self, annotator, route_name='lookup', **process_urn_kwargs):
         """Generate an OPDS feed describing works identified by identifier."""
         urns = flask.request.args.getlist('urn')
 
         this_url = cdn_url_for(route_name, _external=True, urn=urns)
-        response = self.process_urns(urns, **process_urn_kwargs)
-        self.post_lookup_hook()
+        handler = self.process_urns(urns, **process_urn_kwargs)
 
-        if response:
+        if isinstance(handler, ProblemDetail):
             # In a subclass, self.process_urns may return a ProblemDetail
-            return response
+            return handler
 
         opds_feed = LookupAcquisitionFeed(
-            self._db, "Lookup results", this_url, self.works, annotator,
-            precomposed_entries=self.precomposed_entries,
+            self._db, "Lookup results", this_url, handler.works, annotator,
+            precomposed_entries=handler.precomposed_entries,
         )
         return feed_response(opds_feed)
 
+    def process_urns(self, urns, **process_urn_kwargs):
+        """Process a number of URNs by instantiating a URNLookupHandler
+        and having it do the work.
+
+        The information gathered by the URNLookupHandler can be used
+        by the caller to generate an OPDS feed.
+
+        :return: A URNLookupHandler, or a ProblemDetail if
+            there's a problem with the request.
+        """
+        handler = URNLookupHandler(self._db)
+        handler.process_urns(urns, **process_urn_kwargs)
+        return handler
+
     def permalink(self, urn, annotator, route_name='work'):
-        """Look up a single identifier and generate an OPDS feed."""
+        """Look up a single identifier and generate an OPDS feed.
+
+        TODO: This method is tested, but it might not be used and it
+        may be possible to remove it.
```

Comment:
I don't think now is the time to remove this, but I've verified that we could remove it. I think it was used in the content server, but the circulation manager has its own `permalink` implementation and the metadata wrangler doesn't use this.